### PR TITLE
Index v2 for marketplace 

### PR DESCRIPTION
**Pull request by omerKarkKatz**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-5814)

## Description
Create a second zip file called index_v2.zip it will be a copy of index.zip but the README.md images will reference the new server API. 

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/Marketplace/pack_readme_handler.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    """
+    # Open the url image, set stream to True, this will return the stream content.
+    readme_original_url_parsed = urllib.parse.urlparse(readme_original_url)
+    try:
+        logging.info(f'trying to download {readme_original_url_parsed.geturl()}')
+        r = requests.get(readme_original_url_parsed.geturl(), stream=True)
+
+        # Check if the image was retrieved successfully
+        if r.status_code == 200:
+            r.raw.decode_content = True
```
Comment (ilaner) [inline Tests/Marketplace/pack_readme_handler.py:141]: Why this is needed?

File: Tests/private_build/upload_packs_private.py [hunk 2]
```diff
@@ -545,13 +545,36 @@ def main():
     # finished iteration over content packs
     if is_private_build:
         delete_public_packs_from_index(index_folder_path)
-        upload_index_to_storage(index_folder_path, extract_destination_path, private_index_blob, build_number,
-                                private_packs, current_commit_hash, index_generation, is_private_build,
-                                landing_page_sections=landing_page_sections)
+        '''
+        index_folder_path: str, build_number: str, private_packs: list,
```
Comment (ilaner) [inline Tests/private_build/upload_packs_private.py:549]: ?

**Review by yaakovpraisler — CHANGES_REQUESTED** [review]
Nice work!! Most of the comments are about docstrings etc.

- Please go over the functions you changed and update the docstrings accordingly
- Handle the logs you added and leave only what you think is necessary
- Lets also test multiple upload-flow's using the same bucket and make sure the indexs are updated as expected between the flows
- Run also override upload to make sure existing packs with readme urls have changed and that nothing is broken

File: Tests/Marketplace/marketplace_services.py
Comment (DeanArbel) [inline Tests/Marketplace/marketplace_services.py:1]: Please add to the description of the PR that you removed this feature.

File: Tests/Marketplace/pack_readme_handler.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            line = line.replace(url, str(new_replace_url))
+
+            image_gcp_path = Path(gcs_pack_path, BucketUploadFlow.README_IMAGES, image_name)
+            urls_list.append({
+                'original_read_me_url': url,
+                'new_gcs_image_path': image_gcp_path,
+                'image_name': image_name
+            })
+
+        print(line, end='')
```
Comment (gal-forer) [inline Tests/Marketplace/pack_readme_handler.py:98]: Remove print

File: Tests/Marketplace/pack_readme_handler.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+    return urls_list
+
+
+def download_readme_images_from_url_data_list(readme_urls_data_list: list, storage_bucket):
```
Comment (gal-forer) [inline Tests/Marketplace/pack_readme_handler.py:103]: Add annotation for storage_bucket

File: Tests/Marketplace/pack_readme_handler.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    """
+    Iterates over the readme_url_data_list and calls the download_readme_image_from_url_and_upload_to_gcs
```
Comment (gal-forer) [inline Tests/Marketplace/pack_readme_handler.py:105]: Add description for the function arguments

File: Tests/Marketplace/pack_readme_handler.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+            with open(image_name, "rb") as image_file:
+                readme_image.upload_from_file(image_file)
+
+            # remove local saved image
+            os.remove(image_name)
+
+            logging.info(f'Image sucessfully Downloaded: {image_name}')
+            return True
+
+        logging.error(f'Image {image_name} could not be retreived status code {r.status_code}')
```
Comment (gal-forer) [inline Tests/Marketplace/pack_readme_handler.py:157]: There should also be the property `reason` in the response add it as well

File: Tests/Marketplace/pack_readme_handler.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        return False
+    except Exception as e:
+        logging.error(
+            f'Failed downloading the image in url {readme_original_url_parsed}. '
+            f'or failed uploading it to GCP error message {e}')
+        return False
+
+
+def copy_readme_images(production_bucket, build_bucket, images_data: dict, storage_base_path,
```
Comment (gal-forer) [inline Tests/Marketplace/pack_readme_handler.py:166]: Add annotations

**Review by gal-forer — CHANGES_REQUESTED** [review]

File: Tests/Marketplace/marketplace_services.py
Comment (omerKarkKatz) [reply Tests/Marketplace/marketplace_services.py:1]: What feature are you talking about? 
if its all the deleted code from this file. (The functionality remains but moved to different places).

File: Tests/Marketplace/marketplace_services.py [hunk 8]
```diff
@@ -3109,216 +3097,7 @@ def copy_author_image(self, production_bucket, build_bucket, images_data, storag
...
-                logging.error(f'Image {image_name} could not be retreived status code {r.status_code}')
-                return False
-        except Exception as e:
-            logging.error(
-                f'Failed downloading the image in url {readme_original_url}. '
-                f'or failed uploading it to GCP error message {e}')
-            return False
-
-    def upload_images(self, index_folder_path, storage_bucket, storage_base_path, diff_files_list, override_all_packs,
-                      marketplace='xsoar'):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Tests/Marketplace/marketplace_services.py:3321]: Done thanks

File: Tests/Marketplace/marketplace_services.py [hunk 9]
```diff
@@ -3357,8 +3136,6 @@ def upload_images(self, index_folder_path, storage_bucket, storage_base_path, di
             self.cleanup()
             return False
 
-        task_status = self.upload_readme_images(storage_bucket, storage_base_path, diff_files_list, detect_changes,
-                                                marketplace)
         if not task_status:
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Tests/Marketplace/marketplace_services.py:3139]: Done thanks.

File: Tests/Marketplace/pack_readme_handler.py (hunk 3 above)
Comment (omerKarkKatz) [reply Tests/Marketplace/pack_readme_handler.py:98]: This is a print-to-the-file part of the fileinput module. Its intentional.

File: Tests/Marketplace/pack_readme_handler.py (hunk 4 above)
Comment (omerKarkKatz) [reply Tests/Marketplace/pack_readme_handler.py:103]: Storage bucket is part of the google api module and is not annotated in all the upload code.

File: Tests/Marketplace/marketplace_services.py [hunk 10]
```diff
@@ -3109,216 +3097,7 @@ def copy_author_image(self, production_bucket, build_bucket, images_data, storag
...
-        except Exception as e:
-            logging.error(
-                f'Failed downloading the image in url {readme_original_url}. '
-                f'or failed uploading it to GCP error message {e}')
-            return False
-
-    def upload_images(self, index_folder_path, storage_bucket, storage_base_path, diff_files_list, override_all_packs,
-                      marketplace='xsoar'):
+    def upload_images(self, index_folder_path, storage_bucket, storage_base_path, diff_files_list, override_all_packs):
         """
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Tests/Marketplace/marketplace_services.py:3101]: This function was not written as part of this PR, I will add annotation if it is important.

File: Tests/Marketplace/pack_readme_handler.py (hunk 5 above)
Comment (omerKarkKatz) [reply Tests/Marketplace/pack_readme_handler.py:105]: done

File: Tests/Marketplace/pack_readme_handler.py (hunk 7 above)
Comment (omerKarkKatz) [reply Tests/Marketplace/pack_readme_handler.py:166]: done

File: Tests/Marketplace/pack_readme_handler.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+    logging.info('Starting readme images copy.')
+    readme_images = {}
+    if readme_images := images_data.get(BucketUploadFlow.README_IMAGES, None):
+        for pack_name, pack_readme_images_list in readme_images.items():
+            task_status = True
+            err_msg = f"Failed copying {pack_name} pack readme images."
+            pc_uploaded_readme_images = pack_readme_images_list
+
+            if not pc_uploaded_readme_images:
+                logging.info(f"No added/modified readme images were detected in {pack_name} pack.")
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Tests/Marketplace/pack_readme_handler.py:188]: done

File: Tests/private_build/upload_packs_private.py (hunk 2 above)
Comment (omerKarkKatz) [reply Tests/private_build/upload_packs_private.py:549]: done

File: Tests/Marketplace/upload_packs.py [hunk 12]
```diff
@@ -248,32 +248,9 @@ def clean_non_existing_packs(index_folder_path: str, private_packs: list, storag
...
-    :param is_private: Indicates if upload is private.
-    :param force_upload: Indicates if force upload or not.
-    :param previous_commit_hash: The previous commit hash to diff with.
-    :param landing_page_sections: landingPage sections.
-    :param artifacts_dir: The CI artifacts directory to upload the index.json to.
-    :param storage_bucket: The storage bucket object
-    :returns None.
-
-    """
+def prepare_index_json(index_folder_path: str, build_number: str, private_packs: list,
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Tests/Marketplace/upload_packs.py:251]: done

File: Tests/Marketplace/upload_packs.py [hunk 13]
```diff
@@ -301,37 +278,82 @@ def upload_index_to_storage(index_folder_path: str, extract_destination_path: st
...
+    """
+    Upload updated index zip to cloud storage.
+
+    :param index_folder_path: index folder full path.
+    :param extract_destination_path
+    :param index_blob: google cloud storage object that represents index.zip blob.
+    :param index_generation: downloaded index generation.
+    :param is_private: Indicates if upload is private.
+    :param artifacts_dir: The CI artifacts directory to upload the index.json to.
+    :param storage_bucket: The storage bucket object
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply Tests/Marketplace/upload_packs.py:299]: fixed

File: Tests/Marketplace/upload_packs.py [hunk 14]
```diff
@@ -1269,14 +1326,43 @@ def main():
...
 
+    prepare_index_json(index_folder_path=index_folder_path,
+                       build_number=build_number,
+                       private_packs=private_packs,
+                       current_commit_hash=current_commit_hash,
+                       force_upload=force_upload,
+                       previous_commit_hash=previous_commit_hash,
+                       landing_page_sections=statistics_handler.landing_page_sections)
+
+    logging.info('Strting new code index_v2')
```
Comment (yaakovpraisler) [inline Tests/Marketplace/upload_packs.py:1337]:
```suggestion
    logging.info('Starting initialize index_v2')
```

File: Tests/Marketplace/upload_packs.py [hunk 15]
```diff
@@ -1269,14 +1326,43 @@ def main():
...
+    index_v2_local_path, index_v2_blob = init_index_v2(storage_base_path, extract_destination_path,
+                                                       storage_bucket, index_folder_path)
+
+    logging.info('replacing the urls in index_V2')
```
Comment (yaakovpraisler) [inline Tests/Marketplace/upload_packs.py:1341]:
```suggestion
    logging.info('Starting to replace the readme images urls in index_V2')
```

File: Tests/Marketplace/upload_packs.py [hunk 16]
```diff
@@ -1269,14 +1326,43 @@ def main():
...
+                            )
+
+    logging.info('uploading new index')
+
+    upload_index_v2(index_folder_path=index_v2_local_path,
+                    extract_destination_path=extract_destination_path,
+                    index_blob=index_v2_blob,
+                    index_name=GCPConfig.INDEX_V2_NAME)
+
+    logging.info('finished uploading new index')
```
Comment (yaakovpraisler) [inline Tests/Marketplace/upload_packs.py:1365]:
```suggestion
    logging.info('Staring to upload index v2')

    upload_index_v2(index_folder_path=index_v2_local_path,
                    extract_destination_path=extract_destination_path,
                    index_blob=index_v2_blob,
                    index_name=GCPConfig.INDEX_V2_NAME)

    logging.info('Finished uploading index v2')
```

**Review by yaakovpraisler — APPROVED** [review]
Great Work! Just a few logging messages improvments -
Please attach here the links to the test upload builds.

**Comment by omerKarkKatz** [comment]
https://code.pan.run/xsoar/content/-/pipelines/4932617 - A link to a test of the upload flow with the sbp param.

**Pull request merged**